pymc3/step_methods/mlda.py

```diff
@@ -0,0 +1,418 @@
+        tuning will be activated for the first 200 steps.
+    base_tune_interval : int
+        The frequency of tuning for the base proposal. Defaults to 100
+        iterations.
+    model : PyMC Model
+        Optional model for sampling step. Defaults to None
+        (taken from context). This model should be the finest of all
+        multilevel models.
+    mode :  string or `Mode` instance.
+        Compilation mode passed to Theano functions
+    subsampling_rates : integer or list of integers
+        One interger for all levels or a list with one number for each level
+        (excluding the finest level).
+        This is the number of samples generated in level l-1 to propose a sample
+        for level l for all l levels (excluding the finest level). The length of
+        the list needs to be the same as the length of coarse_models.
+    base_blocked : bool
+        To flag to choose whether base sampler (level=0) is a
+        Compound MetropolisMLDA step (base_blocked=False)
+        or a blocked MetropolisMLDA step (base_blocked=True).
+
+    Examples
+    ----------
+    .. code:: ipython
+
+        >>> import pymc3 as pm
+        ... datum = 1
+        ...
+        ... with pm.Model() as coarse_model:
+        ...     x = Normal("x", mu=0, sigma=10)
+        ...     y = Normal("y", mu=x, sigma=1, observed=datum - 0.1)
+        ...
+        ... with pm.Model():
+        ...     x = Normal("x", mu=0, sigma=10)
+        ...     y = Normal("y", mu=x, sigma=1, observed=datum)
+        ...     step_method = pm.MLDA(coarse_models=[coarse_model]
+        ...                           subsampling_rates=5)
+        ...     trace = pm.sample(ndraws=500, chains=2,
+        ...                       tune=100, step=step_method,
+        ...                       random_seed=123)
+        ...
+        ... pm.summary(trace)
+            mean     sd	     hpd_3%	   hpd_97%
+        x	1.011	 0.975	 -0.925	   2.824
+
+    References
+    ----------
+    .. [Dodwell2019] Dodwell, Tim & Ketelsen, Chris & Scheichl,
+    Robert & Teckentrup, Aretha. (2019).
+    Multilevel Markov Chain Monte Carlo.
+    SIAM Review. 61. 509-545.
+        `link <https://doi.org/10.1137/19M126966X>`__
+    """
+
+    name = "mlda"
+
+    # All levels use block sampling,
+    # except level 0 where the user can choose
+    default_blocked = True
+    generates_stats = True
+
+    # These stats are extended within __init__
+    stats_dtypes = [
+        {
+            "accept": np.float64,
+            "accepted": np.bool,
+            "tune": np.bool,
+            "base_scaling": object,
+        }
+    ]
+
+    def __init__(
+        self,
+        coarse_models,
+        vars=None,
+        base_S=None,
+        base_proposal_dist=None,
+        base_scaling=1.0,
+        tune=True,
+        base_tune_interval=100,
+        model=None,
+        mode=None,
+        subsampling_rates=5,
+        base_blocked=False,
+        **kwargs,
```

Comment:
We started doing type annotations with `typing`.  No need to do it for very weakly typed kwargs, if properly described in the docstrings, but I found it very useful to better understand internals.

Reply:
I added typing in mlda.py. I tried to make it as specific as possible for most arguments. 

I am not sure what types to add to things like the inputs/return of astep() as I am not sure if and how these change with different backends.

Also the argument `S` that is fed to the proposals - I was not sure what types are permissible there. 

Let me know if you want me to change anything.